syntax = "proto3";

package NullSpaceIPC;

message ClientIdentificationEvent {
	uint32 dll_major = 1;
	uint32 dll_minor = 2;
	uint32 dll_patch = 3;
	string app_name = 4;
}
message RegionList {
	repeated uint32 regions = 1;
}

message NodeList {
	repeated uint32 nodes = 1;
}

message Location {
	oneof where {
		RegionList regions = 1;
		NodeList nodes = 2;
	}
}

message SimpleHaptic {
	uint32 effect = 1;
	uint32 repetitions = 2;
	float strength = 3;
}

message ContinuousHaptic {
	float strength = 1;
}


message BufferedHaptic {
	repeated float samples = 1;
	float frequency = 2;
}


message BeginAnalogAudio {

}

message EndAnalogAudio {

}

message PlaybackEvent {
	enum Command {
		UNKNOWN = 0;
		PAUSE = 1;
		UNPAUSE = 2;
		CANCEL = 3;
	}
	Command command = 1;
	
}

message LocationalEvent {
	Location location = 1;
	oneof events {
		SimpleHaptic simple_haptic = 2;
		ContinuousHaptic continuous_haptic = 3;
		BufferedHaptic buffered_haptic = 4;
		BeginAnalogAudio begin_analog_audio = 16;
		EndAnalogAudio end_analog_audio = 17;

	}
}

message EnableTracking {

}

message DisableTracking {

}
message DeviceEvent {
	uint32 device = 1;
	oneof events {
		EnableTracking enable_tracking = 16;
		DisableTracking disable_tracking = 17;
	}
}
message HighLevelEvent {
	uint64 parent_id = 1;
	oneof events {
		PlaybackEvent playback_event = 2;
		LocationalEvent locational_event = 3;
		DeviceEvent device_event = 4;
		ClientIdentificationEvent client_id_event = 16;
	}
}
